File: Data Structure/CircularQueue.java
public class CircularQueue {
    int front,rear;
    int queue[];
    int capacity;   
    CircularQueue(int capacity)
    {
        front=rear=-1;
        this.capacity=capacity;
        queue=new int[capacity];
    }
    public void enQueue(int data)
    {
        if((rear+1)%capacity==front)
        {
            System.out.println("Sorry! Queue overflow");
        }
        else 
        {
            rear=(rear+1)%capacity;
            queue[rear]=data;
            System.out.println("New Node Enqueued");
            if(front==-1)
            {
                front=rear;
            }
        }
    }
    public void deQueue()
    {
        if(front==-1)
        {
            System.out.println("Sorry! Queue underflow");
        }
        else 
        {
            int data=queue[front];
            front=(front+1)%capacity;
            System.out.println("Dequeued Node is "+data);
            if(front==rear)
            {
                front=rear=-1;
            }
        }
    }
    public void traverse()
    {
        if(front==-1)
        {
            System.out.println("Sorry! Queue underflow");
        }
        else 
        {
            int i=front;
            while(i!=rear)
            {
                System.out.print(queue[i]+" ");
                i=(i+1)%capacity;
            }
            System.out.println(queue[i]);
        }
    }
    public static void main(String[] args) {
        CircularQueue queue=new CircularQueue(5);
        queue.enQueue(10);
        queue.enQueue(20);
        queue.enQueue(30);
        queue.enQueue(40);
        queue.enQueue(50);
        queue.traverse();
        queue.deQueue();
        queue.traverse();
        queue.deQueue();
        queue.traverse();
        queue.deQueue();
        queue.traverse();
    }
}

File: Data Structure/CircularQueue.c
#include<stdio.h>
#include<stdlib.h>
int capacity;
int front=-1;
int rear=-1;
void enqueue(int * cirQueue)
{
    int item;
    if((rear+1)%capacity==front)
    {
        printf("\nQueue overflow");
    }
    else
    {
        printf("\nEnter Queue item:");
        scanf("%d",&item);
        rear=(rear+1)%capacity;
        cirQueue[rear]=item;
        if(front==-1)
        {
            front++;
        }
        printf("\nItem sucessfully Enqueued.");

    }
}
void dequeue(int * cirQueue)
{
    int temp;
    if(front==-1)
    {
        printf("\nQueue underflow");
    }
    else
    temp=cirQueue[front];
    printf("\nItem dequeued Sucessfully.");
    printf("\nDequeued item:%d",temp);
    if(front==rear)
    {
        front=rear=-1;
    }
    else
    {
        front=(front+1)%capacity;
    }
}
void traverse(int * cirQueue)
{
    int temp;
    if(front==-1)
    {
        printf("Sorry!Queue empty");
    }
    else 
    {
        printf("Queue elements are: ");
        if(rear<front)
        {
            for(temp=front;temp<capacity;temp++)
            {
                printf("%d ",cirQueue[temp]);
            }
            for(temp=0;temp<=rear;temp++)
            {
                printf("%d ",cirQueue[temp]);
            }
        }
        else 
        {
            for(temp=front;temp<=rear;temp++)
            {
                printf("%d ",cirQueue[temp]);
            }
        }
    }
}
void queueStatus(int * cirQueue)
{
    int status;
    if(front==-1)
    {
        printf("\nQueue empty");
    }
    else
    {
        status=(capacity-front+rear)%capacity+1;
        printf("\nQueue status:%d",status);
    }
}
void main()
{
    int choice;
    int cirQueue[20];
    printf("Enter capacity:");
    scanf("%d",&capacity);
    while(1)
    {
        printf("\nPress\n1.enqueue\n2.dequeue\n3.Queue Traverse\n4.Queue Status\n5.exit\n");
        scanf("%d",&choice);
        switch(choice)
            {
                case 1:enqueue(cirQueue);
                        break;
                case 2:dequeue(cirQueue);
                        break;
                case 3:traverse(cirQueue);
                        break;
                case 4:queueStatus(cirQueue);
                        break;
                case 5:exit(0);
                default:printf("Invalid Choice");
            }
    }
}